internal/lang/funcs/filesystem_test.go

```diff
@@ -217,6 +217,63 @@ func TestTemplateFile(t *testing.T) {
 	}
 }
 
+func TestTemplateString(t *testing.T) {
+	tests := []struct {
+		Data  cty.Value
+		Vars  cty.Value
+		Want  cty.Value
+		Err   string
+	}{
+		{
+			cty.StringVal("Hello, Jodie!"),
+			cty.EmptyObjectVal,
+			cty.StringVal("Hello, Jodie!"),
+			``,
+		},
+		{
+			cty.StringVal("Hello, ${name}!"),
+			cty.MapVal(map[string]cty.Value{
+				"name": cty.StringVal("Jodie"),
+			}),
+			cty.StringVal("Hello, Jodie!"),
+			``,
+		},
+		// Add more test cases for templatestring here
+	}
+
+	templateStringFn := MakeTemplateFunc(".", func() map[string]function.Function {
+		return map[string]function.Function{}
+	}, false)
+
+	for _, test := range tests {
+		t.Run(fmt.Sprintf("TemplateString(%#v, %#v)", test.Data, test.Vars), func(t *testing.T) {
+			got, err := templateStringFn.Call([]cty.Value{test.Data, test.Vars})
+
+			if argErr, ok := err.(function.ArgError); ok {
+				if argErr.Index < 0 || argErr.Index > 1 {
+					t.Errorf("ArgError index %d is out of range for templatestring (must be 0 or 1)", argErr.Index)
+				}
+			}
+
+			if test.Err != "" {
+				if err == nil {
+					t.Fatal("succeeded; want error")
+				}
+				if got, want := err.Error(), test.Err; got != want {
+					t.Errorf("wrong error\ngot:  %s\nwant: %s", got, want)
+				}
+				return
+			} else if err != nil {
+				t.Fatalf("unexpected error: %s", err)
+			}
+
+			if !got.RawEquals(test.Want) {
+				t.Errorf("wrong result\ngot:  %#v\nwant: %#v", got, test.Want)
+			}
```

Comment:
I wonder if we can re-rder some of this logic so we don't require the early `return` on line 265. 

Maybe inverting some of the ifs to handle the shorter path first may be favourable here.

Reply:
commit [4fb1950](https://github.com/opentofu/opentofu/pull/1223/commits/4fb1950bb377f66c61af3af40afb85f8d11d343a) resolves this